stdlib/public/core/Array.swift

```diff
@@ -1498,6 +1519,63 @@ extension Array {
     it._position = endIndex
     return (it,buffer.index(buffer.startIndex, offsetBy: self.count))
   }
+
+  /// Calls the given closure with a pointer to the full capacity of the
+  /// array's mutable contiguous storage.
+  ///
+  /// - Parameter body: A closure that can modify or deinitialize existing
+  ///   elements of initialize new elements.
+  ///   - Parameters:
+  ///     - buffer: An unsafe mutable buffer of the array's full storage,
+  ///       including any uninitialized capacity after the initialized
+  ///       elements. Only the elements in `buffer[0..<initializedCount]` are
+  ///       initialized.
+  ///     - initializedCount: The count of the array's initialized elements.
+  ///       If you initialize or deinitialize any elements inside `body`,
+  ///       update `initializedCount` with the new count for the array.
+  /// - Returns: The return value, if any, of the `body` closure parameter.
+  public mutating func withFullCapacityUnsafeMutableBufferPointer<R>(
+    _ body: (
+      _ buffer: inout UnsafeMutableBufferPointer<Element>,
+      _ initializedCount: inout Int
+    ) throws -> R
+  ) rethrows -> R {
+    var initializedCount = self.count
+    let capacity = self.capacity
+
+    // Ensure unique storage
+    reserveCapacity(capacity)
+
+    // Ensure that body can't invalidate the storage or its bounds by
+    // moving self into a temporary working array.
+    // See further notes above in withUnsafeMutableBuffer
+
+    var work = Array()
+    (work, self) = (self, work)
+
+    // Create an UnsafeBufferPointer over the full capacity of `work`
+    // that we can pass to `body`.
+    let pointer = work._buffer.firstElementAddress
+    var inoutBufferPointer = UnsafeMutableBufferPointer(
+      start: pointer, count: capacity)
+
+    // Put the working array back before returning and update the count.
+    defer {
+      _precondition(
+        inoutBufferPointer.baseAddress == pointer &&
+          inoutBufferPointer.count == capacity,
+        """
+        Array withFullCapacityUnsafeMutableBufferPointer: \
+        replacing the buffer is not allowed
```

Comment:
Then why is it being passed `inout`?

Reply:
It looks like I'm cargo-culting the rationale for the buffer being `inout` in `withUnsafeMutableBufferPointer`, which is a *very* old behavior (introduced in c5c17e2615a066452260780c8f6a54e519d54a8a). It looks like that was probably because the type used to access the elements then (`UnsafeMutableArray`) needed to be mutable to access its subscript setter, unlike `UnsafeMutableBufferPointer`.

Reply:
This should probably prefer the correct semantics (non-`inout`) over matching the existing method, esp. since you can't pass the same closure to both anyway. Thanks!

Reply:
Ahhh right, you do need mutation if you want to do collectiony things on the UnsafeMutableBufferPointer. But in this case it's not safe to do collectiony things unless you take a slice anyway, at which point you can use `var`.